lib/srv/app/session.go

```diff
@@ -62,15 +67,35 @@ type sessionChunk struct {
 	fwd *forward.Forwarder
 	// streamWriter can emit events to the audit log.
 	streamWriter events.StreamWriter
+
+	// inflightCond protects and signals change of inflight
+	inflightCond *sync.Cond
+	// inflight is the amount of in-flight requests
+	// closing the chunk is only allowed when this is 0.
+	// On session expiration, this will first be atomically decremented to -1,
+	// preventing any new requests from using the closing/closed session.
+	inflight int64
+	// closeTimeout is the timeout after which close() will forcibly close the chunk,
+	// even if there are still ongoing requests.
+	// E.g. with 5 minute chunk TTL and 2 minute closeTimeout, the chunk will live
+	// for ~7 minutes at most.
+	closeTimeout time.Duration
+
+	log *logrus.Entry
 }
 
 // newSessionChunk creates a new chunk session.
 func (s *Server) newSessionChunk(ctx context.Context, identity *tlsca.Identity, app types.Application) (*sessionChunk, error) {
```

Comment:
This should return a `sessionChunk` with an `inflight` of 1, so we can use it immediately without having to `acquire` it, IMO.